module/mempool/queue/chunkDataPack.go

```diff
@@ -0,0 +1,120 @@
+package queue
+
+import (
+	"sync"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/mempool"
+	herocache "github.com/onflow/flow-go/module/mempool/herocache/backdata"
+	"github.com/onflow/flow-go/module/mempool/herocache/backdata/heropool"
+)
+
+// ChunkDataPackRequestQueue implements a HeroCache-based in-memory queue for storing chunk data pack requests.
+// It is designed to be utilized at Execution Nodes to maintain and respond chunk data pack requests.
+type ChunkDataPackRequestQueue struct {
+	mu        sync.RWMutex
+	cache     *herocache.Cache
+	sizeLimit uint
+}
+
+var _ mempool.ChunkDataPackRequestQueue = &ChunkDataPackRequestQueue{}
+
+func NewChunkDataPackRequestQueue(sizeLimit uint32, logger zerolog.Logger, collector module.HeroCacheMetrics) *ChunkDataPackRequestQueue {
+	return &ChunkDataPackRequestQueue{
+		cache: herocache.NewCache(
+			sizeLimit,
+			herocache.DefaultOversizeFactor,
+			heropool.LRUEjection,
+			logger.With().Str("mempool", "chunk-data-pack-request-queue-herocache").Logger(),
+			collector),
+		sizeLimit: uint(sizeLimit),
+	}
+}
+
+// Push stores chunk data pack request into the queue.
+// Boolean returned variable determines whether push was successful, i.e.,
+// push may be dropped if queue is full or already exists.
+func (c *ChunkDataPackRequestQueue) Push(chunkId flow.Identifier, requesterId flow.Identifier) bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.cache.Size() >= c.sizeLimit {
+		return false
+	}
+
+	req := chunkDataPackRequestEntity{
+		ChunkDataPackRequest: mempool.ChunkDataPackRequest{
+			ChunkId:     chunkId,
+			RequesterId: requesterId,
+		},
+		id: identifierOf(chunkId, requesterId),
+	}
```

Comment:
nit: you can likely create this object outside of the lock

Reply:
Does not seem efficient: 

Moving `id` creation out of critical section like below. Then we generate `id` even if the cache is full and we literally never get to add the `req` to the cache. `id` creation involves encoding and hashing the `chunkId, requesterId` pair and by moving it out of critical section, we pay an extra encoding + hashing cost each time we _try_ adding request to the cache, even if the cache is full and the add transaction aborted. 
```
	id := identifierOf(chunkId, requesterId)
	
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.cache.Size() >= c.sizeLimit {
		return false
	}

	req := chunkDataPackRequestEntity{
		ChunkDataPackRequest: mempool.ChunkDataPackRequest{
			ChunkId:     chunkId,
			RequesterId: requesterId,
		},
		id: id,
	}

	return c.cache.Add(req.id, req)
```

Please note that `cache` is not concurrency-safe, hence `c.cache.Add(req.id, req)` is a critical section that needs to be covered by a lock. The other alternative is to lock/unlock twice, which is more expensive, and also breaks the isolation of the `Push` operation as two duplicate push operations can reach the `point-2` line concurrently while the first one is succeeding and the other one is failing (due to the duplicate one just added). However, for passing from `point-1` to `point-2` in the code, we pay an extra lock acquisition as well as identifier computation for the duplicate one. 
```
        // point-1
	c.mu.Lock()
	if c.cache.Size() >= c.sizeLimit {
		return false
	}

	req := chunkDataPackRequestEntity{
		ChunkDataPackRequest: mempool.ChunkDataPackRequest{
			ChunkId:     chunkId,
			RequesterId: requesterId,
		},
		id: identifierOf(chunkId, requesterId),
	}
	defer c.mu.Unlock()
	
	// point-2
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.cache.Add(req.id, req)
```